sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsPartitions.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.connector.catalog;
+
+import java.util.Map;
+
+import org.apache.spark.annotation.Experimental;
+
+/**
+ * Catalog methods for working with Partitions.
+ */
+@Experimental
+public interface SupportsPartitions extends TableCatalog {
+
+    /**
+     * Create partitions in an existing table, assuming it exists.
+     *
+     * @param ident a table identifier
+     * @param partitions transforms to use for partitioning data in the table
+     * @param ignoreIfExists
+     */
+    void createPartitions(
+            Identifier ident,
+            TablePartition[] partitions,
+            Boolean ignoreIfExists);
+
+    /**
+     * Drop partitions from a table, assuming they exist.
+     *
+     * @param ident a table identifier
+     * @param partitions a list of string map for existing partitions
+     * @param ignoreIfNotExists
+     */
+    void dropPartitions(
+            Identifier ident,
+            Map<String, String>[] partitions,
+            Boolean ignoreIfNotExists);
+
+    /**
+     * Override the specs of one or many existing table partitions, assuming they exist.
+     *
+     * @param ident a table identifier
+     * @param oldpartitions a list of string map for existing partitions to be renamed
+     * @param newPartitions a list of string map for new partitions
+     */
+    void renamePartitions(
+            Identifier ident,
+            Map<String, String>[] oldpartitions,
+            Map<String, String>[] newPartitions);
+
+    /**
+     * Alter one or many table partitions whose specs that match those specified in `parts`,
+     * assuming the partitions exist.
+     *
+     * @param ident a table identifier
+     * @param partitions transforms to use for partitioning data in the table
+     */
+    void alterPartitions(
+            Identifier ident,
+            TablePartition[] partitions);
```

Comment:
What does this do?

`TablePartition` looks like a map of key/value pairs that identify a partition, plus a set of parameters. Does this set metadata for a partition? It would help to know how you intend to use this for Hive tables.

Reply:
yes. This API is mainly used for `HiveTable`, especially for external tables. It is rarely used in other source.
It can be moved it we defined to support `HiveTable` in other branch or pr.

Reply:
Okay, a couple of questions, then. First, does this need to be batched? Is there a Hive MetaStore call that this should correspond to to avoid too many RPCs? If not, then it would be better to alter a single partition at a time, right?

Second, what changing this to `replacePartitionMetadata(identifier, properties)`? That is more clear about what "alter" means. With just the verb "alter", it isn't clear whether the properties are supposed to replace or add to existing properties.

Reply:
This should probably throw `NoSuchPartitionException` as well. But if this is a batch partition operation, then it is more difficult because there could be a mix of partitions that exist and those that do not.

Reply:
> First, does this need to be batched? Is there a Hive MetaStore call that this should correspond to to avoid too many RPCs? If not, then it would be better to alter a single partition at a time, right?

Yes, this API mostly used in batch. Besides, all of the `Partition API`s should work with a single partition, except for `listPartitions` or `listPartitionIdentifiers`.

> Second, what changing this to `replacePartitionMetadata(identifier, properties)`? That is more clear about what "alter" means. With just the verb "alter", it isn't clear whether the properties are supposed to replace or add to existing properties.

Yeap, it's better than `alterPartition`.

Reply:
Looks good.